x-pack/legacy/plugins/apm/public/components/app/TransactionDetails/WaterfallWithSummmary/WaterfallContainer/Waterfall/waterfall_helpers/waterfall_helpers.ts

```diff
@@ -216,79 +216,85 @@ function getDuration(items: IWaterfallItem[]) {
   return timestampEnd - timestampStart;
 }
 
-function createGetTransactionById(itemsById: IWaterfallIndex) {
-  return (id?: IWaterfallItem['id']) => {
-    if (!id) {
-      return undefined;
+const transformTraceItems = ({
+  trace,
+  errorsPerTransaction
+}: TraceAPIResponse) =>
+  trace.items.map(traceItem => {
+    const docType = traceItem.processor.event;
+    switch (docType) {
+      case 'span':
+        return getSpanItem(traceItem as Span);
+      case 'transaction':
+        return getTransactionItem(
+          traceItem as Transaction,
+          errorsPerTransaction
+        );
     }
+  });
 
-    const item = itemsById[id];
-    const isTransaction = item?.docType === 'transaction';
-    if (isTransaction) {
-      return (item as IWaterfallItemTransaction).transaction;
-    }
-  };
-}
+const findWaterfallTransactionById = (
+  waterfallItems: Array<IWaterfallItemSpan | IWaterfallItemTransaction>,
+  id?: IWaterfallItem['id']
+): IWaterfallItemTransaction | undefined =>
+  waterfallItems.find(
+    waterfallItem =>
+      waterfallItem.docType === 'transaction' && waterfallItem.id === id
+  ) as IWaterfallItemTransaction;
+
+const sortWaterfallByStartTime = (
+  waterfallItems: Array<IWaterfallItemSpan | IWaterfallItemTransaction>
+) => {
+  const items = sortBy(waterfallItems, 'start');
+
+  return items.map(item => ({ ...item, offset: item.start || 0 }));
+};
 
 export function getWaterfall(
   { trace, errorsPerTransaction }: TraceAPIResponse,
   entryTransactionId?: Transaction['transaction']['id']
 ): IWaterfall {
   if (isEmpty(trace.items) || !entryTransactionId) {
     return {
-      services: [],
+      serviceColors: {},
       duration: 0,
-      orderedItems: [],
-      itemsById: {},
-      getTransactionById: () => undefined,
-      errorCountByTransactionId: errorsPerTransaction,
-      serviceColors: {}
+      errorsPerTransaction,
+      items: [],
+      getTransactionById: () => undefined
     };
   }
 
-  const waterfallItems = trace.items.map(traceItem => {
-    const docType = traceItem.processor.event;
-    switch (docType) {
-      case 'span':
-        return getSpanItem(traceItem as Span);
-      case 'transaction':
-        return getTransactionItem(
-          traceItem as Transaction,
-          errorsPerTransaction
-        );
-    }
-  });
+  const waterfallItems = transformTraceItems({ trace, errorsPerTransaction });
```

Comment:
`transformTraceItems` is very generic. I think `getWaterfallItems` is slightly better. But if you can find something more precise than "waterfall item" I'd be very happy.

That being said, i thought about dropping this entirely and just having regular spans and transactions. Problem with that is that we'll need some custom fields like `skew` and `offset`. And instead of mutating the span and transaction docs it seemed better to create a new "thing" which became `waterfallItem`.